source-academy/frontend
# Source Academy Contest Voting Update - Implementation of Voting Tierlist

### Description
1. Dependencies required: Backend needs to be running and conform to latest commit so that putting multiple entries in the same tier can be accepted.
2. Summary: This change converts the previous scoring of entries using numeric-based ranking into an interactive tierlist with a graphical user interface.
3. Motivation: The old contest voting system felt like a chore to handle because of the large number of scores needed to be inputted. It also used a ranking-based system and did not accept duplicate score, leading to an entry's score being heavily determined by the other 9 entries it was grouped with for voting, instead of its own merit. In addition, the GUI for voting looked dull and uninspired. Hence, we have changed the voting system to a fairer tier-based voting system, where users can assign each entry to any tier they want without any restrictions. The reduced number of tiers (instead of scores) also simplifies the voting process and makes the GUI more attractive to users. Furthermore, due to the removal of checks on duplicate entries and score boundaries, the numeric-based weightage of the tiers can be altered by anyone seeking to improve the system in the future easily, without compromising security as voters will not be able to enter invalid scores at all. We will be giving users a choice of 5 tiers, S, A, B, C and D to choose from, with the possibility of more tiers such as F/SS in the future.

### Type of change


### How to test
Step 1. Run local version of Source Academy. A course containing at least 2 students and 1 staff is required. At least 11 students is recommended but not compulsory. Both a working frontend and backend are required for this, including a backend version after our other pull request (https://github.com/source-academy/backend/pull/938).
Step 2. Download and extract the Contest Voting test files attached. (Note: You can also use your own contest test files).
Step 3. Access ground control as a staff user and upload the contest-C3.xml file to create and publish a new contest ("Beautiful Runes").
Step 4. Generate contest entry submissions by logging in as the student users and accessing the contest. Minimum of 2 submissions needed.
Step 5. Access ground control again as the staff user and upload the contest-C3-voting.xml file to create and publish a contest voting assignment for the contest.
Step 6. Access as a student user. If n submissions were made in step 4, you should be able to view min(10, n - 1) submissions in the tierlist. Clicking on an entry should enable you to view and run its code on the left hand panel. Dragging and dropping entries should produce the message "Please assign each entry to a tier.", until all entries have been dropped in a tier. Upon all entries being assigned, the message "Saved!" should appear.
Step 7. Return to Academy and finalise the submission. 
Step 8. Next, access the backend database. Run "SELECT * FROM submission_votes ORDER BY updated_at DESC;" You should be able to see your votes with the tier mapped to the corresponding score: S: 10, A: 7, B: 4, C: 2, D: 1.
Step 9. You can make more voting submissions with other student users if desired. Wait for the contest-rolling leaderboard to be updated in the backend. (This happens every 2 hours; however this process can be accelerated by changing the scheduler configuration in /backend/config/config.exs)
Step 10. Log in as a staff user and access the voting assignment. You should be able to see the leaderboard populated with entries with nonzero scores.
Step 11. Access the backend database again and run "SELECT submission_id, relative_score FROM answers WHERE relative_score<>0 ORDER BY updated_at DESC;" You should be able to see the scores corresponding to the leaderboard scores.

[Contest Voting Test Files.zip](https://github.com/source-academy/frontend/files/11169916/Contest.Voting.Test.Files.zip)


### Checklist
